### ISSUE TITLE
`aminmax` will trigger INTERNAL ASSERT if input is empty on cuda

### ISSUE THREAD

**Issue by payphon** [pytorch/pytorch#85439]
### 🐛 Describe the bug

`aminmax` will trigger INTERNAL ASSERT if input is empty on cuda

```py
import torch

def fn(input):
    return input.aminmax(dim=0)

device = "cuda"

input = torch.empty([36, 0], dtype=torch.float32)
input.uniform_(-16, 3)

fn(input.to(device))
# RuntimeError: iter.numel() > 0 && iter.ntensors() - iter.noutputs() == 1 && iter.noutputs() >= 1 INTERNAL ASSERT FAILED at "/opt/conda/conda-bld/pytorch_1663571524876/work/aten/src/ATen/native/cuda/Reduce.cuh":1134, please report a bug to PyTorch.
```

On cpu, it will just return empty tensor

### Versions

torch: '1.13.0.dev20220919'

**Comment by lezcano** [comment]
In fact, I think both the CPU and GPU implementations are wrong (how about that?). I think that, in this case, we should error out, as we do in `amin` and `amax`. See for example:
```python
>>> torch.amin(torch.rand(3,0), dim=1)
IndexError: amin(): Expected reduction dim 1 to have non-zero size.
```

This is because the `amin` and `amax` functions do not have an identity element**.

** In fact they do have one. For `amin` the identity element would be `+inf` for floats and int_max for ints, but we resolved at some point that this would be an unexpected behaviour, so we figured it'd be best to throw an error.

**Comment by lezcano** [comment]
We would accept a PR fixing this issue :)

**Comment by payphon** [comment]
Yes, I think throwing an error is the correct way!

**Comment by nmacchioni** [comment]
> In fact, I think both the CPU and GPU implementations are wrong (how about that?). I think that, in this case, we should error out, as we do in `amin` and `amax`. See for example:
> 
> ```python
> >>> torch.amin(torch.rand(3,0), dim=1)
> IndexError: amin(): Expected reduction dim 1 to have non-zero size.
> ```
> 
> This is because the `amin` and `amax` functions do not have an identity element**.
> 
> ** In fact they do have one. For `amin` the identity element would be `+inf` for floats and int_max for ints, but we resolved at some point that this would be an unexpected behaviour, so we figured it'd be best to throw an error.

From the PyTorch docs it seems that RuntimeError is the appropriate response when the reduced dim has size 0 (https://pytorch.org/docs/stable/generated/torch.aminmax.html), in which case the CPU path needs to be fixed, but I do agree that torch.amin/torch.amax/torch.aminmax should be consistent in the returned error.

**Comment by lezcano** [comment]
Exactly, both the CPU and GPU paths should throw the same error, similar to the one in `amin` and `amax`.

**Comment by ShisuiUzumaki** [comment]
I would like to try my hand on this. Could you guide me on this

**Comment by lezcano** [comment]
A good start would be to `git grep` for the `amin()` error to see how `amin` checks these things, and replicate these in the `aminmax` function. To see what are the possible entry points for the `aminmax` function for the different backends, please have a look at the `native_functions.yaml`. That's the file that binds the Python frontend and the C++ backend.

**Comment by ShisuiUzumaki** [comment]
Okay, I will start with `amin()`. I am assuming that `aminmax()` behaviour should be same as `amin()` in case of empty tensor

**Comment by nmacchioni** [comment]
Yes, aminmax() should return an error similar to that returned from amin() when the reduced dimension is 0-dim.

**Comment by ShisuiUzumaki** [comment]
@lezcano I tried git grep, but I can't seem to understand where `amin()` is throwing the error. It seems as though the error statements in python and C++ are different. After `git grep amin()` what I got is this:
```

aten/src/ATen/native/ReduceOps.cpp:    at::native::zero_numel_check_dims(self, dim, "amin()");
test/test_maskedtensor.py:        _compare_mts(masked_tensor(torch.tensor(-4.0), torch.tensor(True)), mt.amin())
test/test_maskedtensor.py:        mt.amin().backward()
test/test_reductions.py:                 (lambda x: x.amin(), 'amin'),
test/test_reductions.py:        self.assertEqual(x[:, :2].amin().item(), -1)
torch/masked/_ops.py:            # sum()/prod() return the reduction identity when dim has size 0 but amax()/amin() do not
```

I tried looking at ` at::native::zero_numel_check_dims(self, dim, "amin()");` this for more clarity, but it does not seem to help. Could you provide more hints?

**Comment by nmacchioni** [comment]
@ShisuiUzumaki Have you tried looking at `native_functions.yaml` yet? That file dictates which C++ backend files get called in the transfer from Python. If you look carefully, you will see that `aten::native::amin_out` gets called which then checks the dimension in a call to `at::native::zero_numel_check_dims`. What you need to do is find the C++ function that gets called by `aminmax()` and add checks like those in `aten::native::amin_out`.

**Comment by ShisuiUzumaki** [comment]
> @ShisuiUzumaki Have you tried looking at `native_functions.yaml` yet? That file dictates which C++ backend files get called in the transfer from Python. If you look carefully, you will see that `aten::native::amin_out` gets called which then checks the dimension in a call to `at::native::zero_numel_check_dims`. What you need to do is find the C++ function that gets called by `aminmax()` and add checks like those in `aten::native::amin_out`.

Ok, I will follow your advice

**Comment by ShisuiUzumaki** [comment]
> @ShisuiUzumaki Have you tried looking at `native_functions.yaml` yet? That file dictates which C++ backend files get called in the transfer from Python. If you look carefully, you will see that `aten::native::amin_out` gets called which then checks the dimension in a call to `at::native::zero_numel_check_dims`. What you need to do is find the C++ function that gets called by `aminmax()` and add checks like those in `aten::native::amin_out`.

@nmacchioni  `at::native::zero_numel_check_dims` is being called in `TORCH_META_FUNC(amin)`. Reference:

```
TORCH_META_FUNC(amin)
(const Tensor& self, IntArrayRef dim, bool keepdim) {
  auto maybe_result = maybe_get_output();
  if (maybe_result.defined()) {
    TORCH_CHECK(self.scalar_type() == maybe_result.scalar_type(), "Expected the dtype for input and out to match, but got ",
                self.scalar_type(), " for input's dtype and ",  maybe_result.scalar_type(), " for out's dtype.");
  }
  if (self.numel() == 0) {
    

> at::native::zero_numel_check_dims(self, dim, "amin()");

  }
  const ScalarType& out_dtype = maybe_result.defined() ? maybe_result.scalar_type() : self.scalar_type();
  resize_reduction(*this, self, dim, keepdim, out_dtype);
}
```

**Comment by nmacchioni** [comment]
@ShisuiUzumaki Great, now that you've found where the check happens, apply something similar for `aminmax()`

**Comment by ShisuiUzumaki** [comment]
> @ShisuiUzumaki Great, now that you've found where the check happens, apply something similar for `aminmax()`

I found this as well:

```
TORCH_META_FUNC(aminmax)
(const Tensor& self, c10::optional<int64_t> dim_opt, bool keepdim) {
  DimVector shape;
  if (dim_opt.has_value()) {
    auto dim = maybe_wrap_dim(dim_opt.value(), self.ndimension());
    at::native::zero_numel_check_dims(self, dim, "aminmax");
    shape = get_reduction_shape(self, dim, keepdim);
  } else {
    TORCH_CHECK(
        self.numel() > 0,
        "aminmax(): cannot compute aminmax over an empty dimension as the "
        "operation has no identity.");
    if (keepdim) {
      shape = DimVector(self.ndimension(), 1);
    }
  }
  const auto options = self.options();
  this->set_output_raw_strided(0, shape, {}, options);
  this->set_output_raw_strided(1, shape, {}, options);
} 
```

As you can see, in the above code,  `native::zero_numel_check_dims(self, dim, "aminmax");` has no `at` namespace which could be the potential reason for `ASSERT` fail error.

**Comment by lezcano** [comment]
Good news! You have located the correct part of the code. Bad news! My initial suggestion for the fix was wrong. I will write down a comprehensive solution on how to solve this.

In the meantime, you can try building PyTorch from source. If you are going to make changes on the C++ code, you'll need to recompile

**Comment by lezcano** [comment]
So, many functions in PyTorch are split: A metafunc (the one you located) that runs all the shape / device / arg checks and sets the shape and strides of the output, and the implementation itself. The implementation for this function is here:
https://github.com/pytorch/pytorch/blob/ad87365e54e7b20b49ac23ee325f1da732655808/aten/src/ATen/native/ReduceOps.cpp#L316-L338
You see that in the branch that we're interested on we call the "function" `aminmax_stub`. This dispatches to a CPU or CUDA implementation depending on the `device()` that it gets. Doing a `git grep aminmax_stub`, takes us to
https://github.com/pytorch/pytorch/blob/ad87365e54e7b20b49ac23ee325f1da732655808/aten/src/ATen/native/cuda/ReduceOps.cpp#L96
This line means that the kernel that's executed in the CUDA case is called `aminmax_kernel_impl`
https://github.com/pytorch/pytorch/blob/ad87365e54e7b20b49ac23ee325f1da732655808/aten/src/ATen/native/cuda/ReduceOps.cpp#L62-L67
That kernel creates a `TensorIterator`, which is a helper class to implement elementwise and reduction kernels. Now, it sometimes does not work well when iterating over tensors of zero elements, so you need to put a guard around the execution similar to how `amin` does it:
https://github.com/pytorch/pytorch/blob/ad87365e54e7b20b49ac23ee325f1da732655808/aten/src/ATen/native/ReduceOps.cpp#L1494-L1496

Exercise. Have a look at whether the other branch of `aminmax_out` and their CPU dispatches also fail, and if so, fix them

Then, add one or a few samples here that test these cases:
https://github.com/pytorch/pytorch/blob/ad87365e54e7b20b49ac23ee325f1da732655808/torch/testing/_internal/common_methods_invocations.py#L2553-L2561
and run the tests with `pytest test/test_ops*.py -vk aminmax`

**Comment by ShisuiUzumaki** [comment]
@lezcano   Let me sort this out:

1.  I need to put a guard around `aminmax_launch_kernel(iter); ` 
[pytorch/aten/src/ATen/native/cuda/ReduceOps.cpp](https://github.com/pytorch/pytorch/blob/ad87365e54e7b20b49ac23ee325f1da732655808/aten/src/ATen/native/cuda/ReduceOps.cpp#L66)
3.  I need to check CPU implementation for the same issue
4.  Compile torch again.
5.  Perform test  by running `pytest test/test_ops*.py -vk aminmax`

Could you explain the compilation process? Just to be on the safe side

**Comment by lezcano** [comment]
For how to compile PyTorch from source, see `Contributing.md`.

**Comment by ShisuiUzumaki** [comment]
You mean I have understood rest correctly, right?

**Comment by lezcano** [comment]
The rest is on point, yes :)

**Comment by ShisuiUzumaki** [comment]
Ok, I am on it

**Comment by janeyx99** [comment]
This looks like it's been fixed, closing
